subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java

```diff
@@ -260,97 +256,4 @@ public static Action<ConfigurationInternal> configureLibraryElementsAttributeFor
             }
         };
     }
-
-    /**
-     * A custom artifact type which allows the getFile call to be done lazily only when the
-     * artifact is actually needed.
-     */
-    private abstract static class IntermediateJavaArtifact extends AbstractPublishArtifact {
-        private final String type;
-
-        public IntermediateJavaArtifact(TaskDependencyFactory taskDependencyFactory, String type, Object dependency) {
-            super(taskDependencyFactory, dependency);
-            this.type = type;
-        }
-
-        @Override
-        public String getName() {
-            return getFile().getName();
-        }
-
-        @Override
-        public String getExtension() {
-            return "";
-        }
-
-        @Override
-        public String getType() {
-            return type;
-        }
-
-        @Nullable
-        @Override
-        public String getClassifier() {
-            return null;
-        }
-
-        @Override
-        public Date getDate() {
-            return null;
-        }
-
-        @Override
-        public boolean shouldBePublished() {
-            return false;
-        }
-    }
-
-    /**
-     * An {@link IntermediateJavaArtifact} with a non-lazy File.
-     */
-    public static class ImmediateIntermediateJavaArtifact extends IntermediateJavaArtifact {
-
-        private final File file;
-
-        public ImmediateIntermediateJavaArtifact(TaskDependencyFactory taskDependencyFactory, String type, Object dependency, File file) {
-            super(taskDependencyFactory, type, dependency);
-            this.file = file;
-        }
-
-        @Override
-        public File getFile() {
-            return file;
-        }
-    }
-
-    /**
-     * An {@link IntermediateJavaArtifact} which achieves lazy file access via a {@link Provider} instead
-     * of inheritance.
-     */
-    public static class ProviderBasedIntermediateJavaArtifact extends IntermediateJavaArtifact {
-
-        private final Provider<File> fileProvider;
-
-        // Used in the Gradle build;
-        // TODO: remove once the usage in gradlebuild.test-fixtures.gradle.kts is no longer there
-        /**
-         * @deprecated Use the overload accepting a TaskDependencyFactory
-         */
-        @Deprecated
-        public ProviderBasedIntermediateJavaArtifact(
```

Comment:
This constructor is currently being used by the Gradle build itself: https://github.com/gradle/gradle/blob/b43779201c11f5dcd0c575bfc4daab5dcaf6b400/build-logic/integration-testing/src/main/kotlin/gradlebuild.test-fixtures.gradle.kts#L81

Ideally the build would not be using this internal class. We should have a public solution for the build itself. 

In the current state of this PR, after the merge, once we update the build to the new nightly it will fail since this class is missing. 

If you want to delete this class and migrate the build to the new class we need to do it in steps. 
1) Leave this class where it is. Add the new class
2) Merge the PR. Update the gradle build to use the new internal class
3) Put out a PR removing this class. 

Its better in the long run if we can avoid using internal classes altogether in our build.

Reply:
I found some example code which seems to do something similar to our own build: https://github.com/JetBrains/kotlin/blob/2e7e1625d301fc23b16fdb6c4e5e5f63a7e65bc6/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/Kotlin2JvmSourceSetProcessor.kt#L74-L76

Though, when I try to implement it in our own build files I get errors. I'm guessing its some kind of kotlin language version thing?

This seems to work but, but when I try to simplify it to a lambda I get compiler errors: 
```
    artifact(processResources.map { it.destinationDir }, object: Action<ConfigurablePublishArtifact> {
        override fun execute(t: ConfigurablePublishArtifact) {
            t.type = ArtifactTypeDefinition.JVM_RESOURCES_DIRECTORY
        }
    })
```

Reply:
I actually don't think this is needed anymore, as the ide-play project doesn't exist.  IDK why my refactorings and deletion didn't catch this usage, but I think it can be removed entirely, which will make this a simpler change.